File: multiSeriesV4.R
# The following function access the BLS API and retreives the data for a list of series names.
# If a series does not exists, a message is printed to the screen
# If a series exists the data is being appended to a DataFrame and a message is being printed to the screen

multiSeriesV4 = function(varList,myKey){ 
    
    # Input: varList = a list of strings containing the series names
    # Input: myKey =  a string containing your BLS API key
    # Output: new_df = a data frame containing the data collected
    
    base_url = 'https://api.bls.gov/publicAPI/v2/timeseries/data/'  #this will not change
    headers = c('Content-type' = 'application/json')  #This will not change !

    parameters = list(
        "seriesid" = varList,
        "startyear" = "2017", 
        "endyear" = "2022",
        "catalog" = TRUE, 
        "calculations" = FALSE, 
        "annualaverage" = FALSE,
        "aspects" = FALSE,
        "registrationkey" = myKey
    )
    
    data = toJSON(parameters, pretty=TRUE, auto_unbox = TRUE) #this converts the R list into a JSON format
    
    p = POST(url = base_url, 
         body = data, 
         add_headers(headers))
    
    json_data = fromJSON(rawToChar(p$content), simplifyDataFrame = FALSE)
    
    n = length(varList) #number of series requested
    
    new_df = data.frame('year'=numeric(),'period'=character())
    for(i in 1:n){
        l = length(json_data$Results$series[[i]]$data) #length of the list 
        if(l == 0){
            print(paste0('Series ',varList[i],' does not exist'))
        } else{
            print(paste0('Series ',varList[i],' exists with ', l , ' observations'))
            d = json_data$Results$series[[i]]$data
            current_df = fromJSON(toJSON(d))
            current_df = current_df[c("year","period","value")]
            current_df$value <- as.numeric(current_df$value)  #convert strings to numbers
            current_df$period <- as.character(current_df$period)
            current_df$year <- as.numeric(current_df$year)
            names(current_df)[names(current_df)=='value'] <- varList[i]
            new_df = merge(new_df, current_df, by = c('year','period'), all = TRUE)
        }
    }       
    return(new_df)
}

File: multiSeriesV4.py
# The following function access the BLS API and retreives the data for a list of series names.
# If a series does not exists, a message is printed to the screen
# If a series exists the data is being appended to a DataFrame and a message is being printed to the screen

# Written by Arie Beresteanu
# This version: February 28, 2022

def multiSeriesV4(varList,myKey):
    # Input: varList = a list of strings containing the series names
    # Input: myKey =  a string containing your BLS API key
    # Output: new_df = a data frame containing the data collected
    
    base_url = 'https://api.bls.gov/publicAPI/v2/timeseries/data/'  #this will not change
    headers = {'Content-type': 'application/json'}  #This will not changed !

    parameters = {
        "seriesid":varList,
        "startyear":"2017", 
        "endyear":"2022",
        "catalog":True, 
        "calculations":False, 
        "annualaverage":False,
        "aspects":False,
        "registrationkey": myKey 
     }

    data = json.dumps(parameters) #this converts the Python dictionary into a JSON format
    
    p = requests.post(base_url, data=data, headers=headers)
    json_data = json.loads(p.text)
    
    n = len(varList) #number of series requested
    
    new_df = pd.DataFrame(columns=['year', 'period'])
    for item in range(n):
        l = len(json_data['Results']['series'][item]['data']) #length of the list
        if l == 0:
            print('Series ',varList[item],' does not exist')
        else:
            
            print('Series ',varList[item],' exists with ',l,' observations')
            d = json_data['Results']['series'][item]['data']
            current_df = pd.DataFrame(data=d)
            current_df = current_df[['year','period','value']].astype({'value': 'float64'})
            current_df = current_df.rename(columns = {'value':varList[item]})
            new_df = new_df.merge(current_df, on = ['year','period'],how='outer')
    
    return new_df
